File: sidecars/hunter_py/rate_limiter.py
import asyncio
import time
import random
import os
import json
from typing import List, Callable

class TokenBucketRateLimiter:
    def __init__(self, capacity: int = 5, refill_rate: int = 15, refill_interval_ms: int = 60_000):
        self.capacity = capacity
        self.tokens = float(capacity)
        self.refill_rate = refill_rate
        self.refill_interval_ms = refill_interval_ms / 1000.0  # Convert to seconds
        self.last_refill = time.time()
        self.current_backoff_s = 0.0
        self.base_backoff_s = 30.0
        self.max_backoff_s = 900.0  # 15 minutes
        self.is_paused_for_captcha = False
        self.pending_requests: List[asyncio.Future] = []
        self._lock = asyncio.Lock()

    def _refill(self):
        now = time.time()
        elapsed = now - self.last_refill
        if elapsed >= self.refill_interval_ms:
            intervals = elapsed // self.refill_interval_ms
            new_tokens = intervals * self.refill_rate
            self.tokens = min(float(self.capacity), self.tokens + new_tokens)
            self.last_refill = now - (elapsed % self.refill_interval_ms)

    def _get_jitter(self, min_s: float = 2.0, max_s: float = 8.0) -> float:
        # Use random.SystemRandom for security/crypto-level randomness
        return random.SystemRandom().uniform(min_s, max_s)

    async def acquire(self):
        if os.getenv("SENTINEL_DEV_BYPASS_RATE_LIMIT") == "true":
            return

        async with self._lock:
            while True:
                if self.is_paused_for_captcha:
                    fut = asyncio.get_event_loop().create_future()
                    self.pending_requests.append(fut)
                    await fut
                    continue

                if self.current_backoff_s > 0:
                    delay = self.current_backoff_s
                    self.current_backoff_s = 0.0
                    await asyncio.sleep(delay)

                self._refill()

                if self.tokens >= 1.0:
                    self.tokens -= 1.0
                    jitter = self._get_jitter()
                    await asyncio.sleep(jitter)
                    return
                else:
                    now = time.time()
                    time_until_refill = self.refill_interval_ms - (now - self.last_refill)
                    await asyncio.sleep(max(0.0, time_until_refill))

    def on_rate_limit(self):
        if self.current_backoff_s == 0:
            self.current_backoff_s = self.base_backoff_s
        else:
            self.current_backoff_s = min(self.current_backoff_s * 2, self.max_backoff_s)
        print(f"[WARN] Rate limit hit. Backing off for {self.current_backoff_s}s")

    def on_captcha(self, portal_id: str):
        self.is_paused_for_captcha = True
        print(json.dumps({
            "event": "hitl_required",
            "type": "captcha",
            "portal_id": portal_id
        }))

    def resume(self):
        self.is_paused_for_captcha = False
        for fut in self.pending_requests:
            if not fut.done():
                fut.set_result(None)
        self.pending_requests = []

File: sidecars/hunter/src/rate-limiter.ts
/* eslint-disable no-console */
import { randomBytes } from 'crypto';

export class TokenBucketRateLimiter {
  private tokens: number;
  private readonly capacity: number;
  private readonly refillRate: number; // Tokens to add
  private readonly refillIntervalMs: number;
  private lastRefill: number;
  private currentBackoffMs: number;
  private readonly baseBackoffMs: number = 30_000;
  private readonly maxBackoffMs: number = 900_000; // 15 minutes
  private isPausedForCaptcha = false;
  private pendingRequests: (() => void)[] = [];

  constructor(capacity = 5, refillRate = 15, refillIntervalMs = 60_000) {
    this.capacity = capacity;
    this.tokens = capacity;
    this.refillRate = refillRate;
    this.refillIntervalMs = refillIntervalMs;
    this.lastRefill = Date.now();
    this.currentBackoffMs = 0;
  }

  /**
   * Refills the bucket based on time passed.
   */
  private refill() {
    const now = Date.now();
    const elapsedTime = now - this.lastRefill;

    if (elapsedTime >= this.refillIntervalMs) {
      // Calculate how many intervals have passed
      const intervals = Math.floor(elapsedTime / this.refillIntervalMs);
      this.tokens = Math.min(this.capacity, this.tokens + intervals * this.refillRate);
      this.lastRefill = now - (elapsedTime % this.refillIntervalMs);
    }
  }

  /**
   * Generates a random delay between min and max using crypto.
   * Do NOT use Math.random() for security reasons per guidelines.
   */
  private getJitter(min = 2000, max = 8000): number {
    const range = max - min;
    const randomBuffer = randomBytes(4);
    const randomNum = randomBuffer.readUInt32LE(0);
    // Max uint32 is 4294967295
    const randomFloat = randomNum / 4294967296;
    return min + Math.floor(randomFloat * range);
  }

  /**
   * Main entry point to acquire a token. Will block until a token is available
   * and any required jitter or backoff has elapsed.
   */
  public async acquire(): Promise<void> {
    // 1. Check Dev Bypass
    if (process.env.SENTINEL_DEV_BYPASS_RATE_LIMIT === 'true') {
      console.warn('[WARN] Rate limit bypassed due to SENTINEL_DEV_BYPASS_RATE_LIMIT=true');
      return;
    }

    // 2. Queue up if paused or out of tokens
    return new Promise<void>((resolve) => {
      const attemptAcquire = async () => {
        if (this.isPausedForCaptcha) {
          this.pendingRequests.push(() => {
            void attemptAcquire();
          });
          return;
        }

        if (this.currentBackoffMs > 0) {
          // Wait for backoff
          const delay = this.currentBackoffMs;
          this.currentBackoffMs = 0; // Reset backoff for next time, it grows on subsequent 429s
          await new Promise((r) => setTimeout(r, delay));
        }

        this.refill();

        if (this.tokens >= 1) {
          this.tokens -= 1;
          const jitter = process.env.NODE_ENV === "test" ? 0 : this.getJitter();
          setTimeout(() => {
            resolve();
          }, jitter);
        } else {
          // Calculate time until next refill and wait
          const now = Date.now();
          const timeUntilRefill = this.refillIntervalMs - (now - this.lastRefill);
          setTimeout(
            () => {
              void attemptAcquire();
            },
            Math.max(0, timeUntilRefill)
          );
        }
      };

      void attemptAcquire();
    });
  }

  /**
   * Triggered on HTTP 429 or CAPTCHA soft-blocks.
   * Starts exponential backoff: 30s -> 60s -> 120s -> ... capped at 15m.
   */
  public onRateLimit() {
    if (this.currentBackoffMs === 0) {
      this.currentBackoffMs = this.baseBackoffMs;
    } else {
      this.currentBackoffMs = Math.min(this.currentBackoffMs * 2, this.maxBackoffMs);
    }
    console.warn(`[WARN] Rate limit hit. Backing off for ${String(this.currentBackoffMs)}ms`);
  }

  /**
   * Triggered on hard CAPTCHA. Pauses all operations until resumed.
   */
  public onCaptcha(portalId: string) {
    this.isPausedForCaptcha = true;
    console.log(
      JSON.stringify({
        event: 'hitl_required',
        type: 'captcha',
        portalId,
      })
    );
  }

  /**
   * Resumes operations after CAPTCHA solved.
   */
  public resume() {
    this.isPausedForCaptcha = false;
    const pending = [...this.pendingRequests];
    this.pendingRequests = [];
    pending.forEach((req) => {
      req();
    });
  }
}
